rfcs/0004-range-queries.md

```diff
@@ -0,0 +1,219 @@
+- [Implementation](#implementation)
+   * [Checkpoint References](#checkpoint-references)
+   * [Memory Usage](#memory-usage)
+- [Rejected Alternatives & Follow-Ups](#rejected-alternatives-follow-ups)
+- [Updates](#updates)
+
+<!-- TOC end -->
+
+Status: Discussion
+
+Authors:
+
+* [Jason Gustafson](https://github.com/hachikuji)
+
+References:
+
+* https://github.com/slatedb/slatedb/issues/8
+* https://github.com/facebook/rocksdb/wiki/Iterator
+* https://github.com/facebook/rocksdb/wiki/Iterator-Implementation
+* https://github.com/slatedb/slatedb/pull/260
+* https://github.com/slatedb/slatedb/pull/278
+
+## Motivation
+
+
+The ability to scan the keys in a range is an essential feature for building applications.
+For example, this can be used to implement a collections of keys using a common prefix. 
+Currently, SlateDB only supports fetching individual values for known keys.
+
+## Goals
+
+The goal of this work is to provide a flexible capability to query a range of keys in the
+database. It should be possible to specify unbounded ranges to allow querying all keys,
+or bounded ranges for finer precision and efficiency.
+
+Range queries have the potential to be expensive from a resource perspective because they
+hold references to existing tables, which prevents garbage collection. Our goal here is
+to  provide a simple initial approach to state management which optimizes for common cases,
+while leaving room for further optimization in the future.
+
+## Public API
+
+We introduce a new `Db::scan` API which accept a `DbRange` defining the range of  the scan.
+The result will be a `DbIterator` object which iterates the current records within the range.
+Users may either provide an explicit starting range in the call to `scan`, or they
+may `seek` using the returned iterator.
+
+```rust
+impl Db {
+    /// Scan a range of keys.
+    /// 
+    /// returns a `DbIterator`  
+    pub async fn scan(
+        &self,
+        range: &DbRange
+    ) -> Result<DbIterator, SlateDbError>;
+}
+
+struct DbRange {
+    lower: Bound<&[u8]>,
+    upper: Bound<&[u8]>,
+}
+
+impl DbIterator {
+    /// Get the next record in the scan.
+    /// 
+    /// returns Ok(None) when the scan is complete
+    /// returns Err(InvalidatedIterator) if the 
+    pub async fn next(
+        &mut self,
+    ) -> Result<Option(DbRecord), SlateDbError>;
+
+    /// Seek to a key ahead of the last key returned from the iterator or
+    /// the lower range bound if no records have yet been returned.
+    /// 
+    /// returns Ok(()) if the position is successfully advanced
+    /// returns SlateDbError::InvalidArgument if `lower_bound` is `Unbounded`
+    /// returns SlateDbError::InvalidArgument if the key is comes before the
+    ///  current iterator position.
+    /// returns Err(InvalidatedIterator) if the 
+    pub async fn seek(
+        &mut self,
+        lower_bound: Bound<&[u8]>
+    ) -> Result<(), SlateDbError>;
+}
+
+struct DbRecord {
+    key: &[u8],
```

Comment:
I suspect we might want `Bytes` for key and value.

Reply:
I agree that makes sense given current signature for `get`. I'll update. My only hesitation would be about exposing third-party structs through the public API, but perhaps `bytes` is standard enough and unlikely to change.

Reply:
Yea.. I think there's a discussion to be had very soon about whether we should be using `AsRef<[u8]>`. This is something @KaguraMilet has pointed out a number of times. I suspect we'll want to do this for the public facing APIs, but in the meantime, I prefer keeping things uniform.